### PR TITLE
Basic redux integration for thread component

### PR DESCRIPTION
**Pull request by DmitryTsepelev**
Here is a very basic integration with redux. I've moved all the complex logic from the Thread component to the reducer, which is also responsible for restoring the initial state from the local storage. Reducer is already covered with a couple of specs, would be also useful to have a spec for getters. Related to #95 and #96

### CODE REVIEW

File: web/app/components/thread/thread.jsx [hunk 1]
```diff
@@ -1,67 +1,41 @@
...
+  render(props) {
     const {
+      collapsed,
       data: { comment, replies = [] },
       mods = {},
     } = props;
 
+    let threadComments = null;
+    if (!collapsed && !!replies.length) {
+      threadComments = replies.map(thread => (
```
Comment (Guria) [inline web/app/components/thread/thread.jsx]: I prefer inline variant. Please put it back to render

File: web/app/components/thread/thread.reducers.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  saveCollapsedComments(collapsedComments);
+};
+
+export const collapsedThreads = (state = initialState, action) => {
+  switch (action.type) {
+    case THREAD_SET_COLLAPSE: {
+      const currentCollapsed = state[action.comment.id];
+
+      if (action.collapsed !== currentCollapsed) {
+        persistCollapsedComments(action.comment, action.collapsed);
```
Comment (Guria) [inline web/app/components/thread/thread.reducers.js]: Reducers must be pure functions. No side effects allowed.

File: web/app/components/thread/thread.jsx [hunk 3]
```diff
@@ -76,26 +50,21 @@ export default class Thread extends Component {
...
-  localStorage.setItem(LS_COLLAPSE_KEY, JSON.stringify(comments));
-}
+const mapDispatchToProps = { setCollapse };
+
+const enhance = connect(
+  mapStateToProps,
+  mapDispatchToProps
+);
+
+export default enhance(Thread);
```
Comment (Guria) [inline web/app/components/thread/thread.jsx]: I also prefer inline declarations here as well:
```js
export default connect(
  (state, props) => ({
    collapsed: getThreadIsCollapsed(state, props.data.comment)
  }),
  { setCollapse }
)(Thread);
```

File: web/app/components/thread/thread.reducers.js (hunk 2 above)
Comment (Guria) [reply web/app/components/thread/thread.reducers.js]: We can make custom middleware to persist to localStorage. Using localStorage value for initial value is ok.

File: web/app/components/thread/thread.jsx (hunk 3 above)
Comment (DmitryTsepelev) [reply web/app/components/thread/thread.jsx]: It will become unreadable as soon as we wrap it up with one more HOC :)

Comment (DmitryTsepelev) [reply web/app/components/thread/thread.jsx]: ✅

File: web/app/components/thread/thread.jsx (hunk 1 above)
Comment (DmitryTsepelev) [reply web/app/components/thread/thread.jsx]: ✅

File: web/app/components/thread/thread.reducers.js (hunk 2 above)
Comment (DmitryTsepelev) [reply web/app/components/thread/thread.reducers.js]: Agreed, done. If there will be more cases like this - we could try handling all the side effects with sagas

**Comment by DmitryTsepelev** [comment]
@Guria Thanks for the review, I think I'm done. Please let me know if we need to change something else

File: web/app/store.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+import { combineReducers } from 'redux';
+
+import { siteId, url } from 'common/settings';
+import { threadReducers, THREAD_SET_COLLAPSE, getCollapsedComments, saveCollapsedComments } from './components/thread';
+
+const rootReducer = combineReducers({
+  ...threadReducers,
+});
+
+const collapsedCommentsMiddleware = ({ getState }) => next => action => {
```
Comment (Guria) [inline web/app/store.js]: maybe export middleware from thread as well?

**Comment by DmitryTsepelev** [comment]
@Guria I'm stuck with the devtools :( I've configured everything according to the docs, aliased all the react-related things, but redux dev tool does not see the store, while react dev tool is "waiting for react" or "waiting for the roots to load", what am I missing here?

**Comment by Guria** [comment]
Ok, nevermind. We can do it in a separate PR. But we need to resolve new conflicts here.

**Comment by DmitryTsepelev** [comment]
Done

**Comment by DmitryTsepelev** [comment]
@Guria Are we good to go? :) I have one more PR to submit

**Comment by Guria** [comment]
@DmitryTsepelev please don't use yarn. only npm 6.1.0+

**Comment by DmitryTsepelev** [comment]
ah ok, got it

**Pull request merged**